helpers/chatHelper.py

```diff
@@ -190,468 +204,613 @@ async def part_channel(
+        client_channel_name = "#multiplayer"
 
-        # Make sure the user is not in restricted mode
-        if osuToken.is_restricted(userToken["privileges"]):
-            raise exceptions.userRestrictedException()
+    return channel_name, client_channel_name
 
-        # Make sure the user is not silenced
-        if await osuToken.isSilenced(userToken["token_id"]):
-            raise exceptions.userSilencedException()
 
-        # Redirect !report to chatbot
-        if message.startswith("!report"):
-            recipient_name = glob.BOT_NAME
-
-        # Determine internal name if needed
-        # (toclient is used clientwise for #multiplayer and #spectator channels)
-        toClient = recipient_name
-        if recipient_name == "#spectator":
-            if userToken["spectating_user_id"] is None:
-                s = userToken["user_id"]
-            else:
-                s = userToken["spectating_user_id"]
-            recipient_name = f"#spect_{s}"
-        elif recipient_name == "#multiplayer":
-            recipient_name = f"#mp_{userToken['match_id']}"
-        elif recipient_name.startswith("#spect_"):
-            toClient = "#spectator"
-        elif recipient_name.startswith("#mp_"):
-            toClient = "#multiplayer"
-
-        isChannel = recipient_name[0] == "#"
-
-        # Make sure the message is valid
-        if not message.strip():
-            raise exceptions.invalidArgumentsException()
-
-        # Truncate really long messages
-        if len(message) > 1024 and userToken["user_id"] != CHATBOT_USER_ID:
-            message = f"{message[:1024]}... (truncated)"
-
-        action_msg = message.startswith("\x01ACTION")
-
-        # Send the message
-        if isChannel:
-            # CHANNEL
-            # Make sure the channel exists
-            channel = await channelList.getChannel(recipient_name)
-            if channel is None:
-                raise exceptions.channelUnknownException()
+async def enqueue_private_message(
+    user_token: osuToken.Token,
+    recipient_token: osuToken.Token,
+    message: str,
+) -> None:
+    packet = serverPackets.sendMessage(
+        fro=user_token["username"],
+        to=recipient_token["username"],
+        message=message,
+        fro_id=user_token["user_id"],
+    )
+    await osuToken.enqueue(recipient_token["token_id"], packet)
 
-            # Make sure the channel is not in moderated mode
-            if channel["moderated"] and not osuToken.is_staff(userToken["privileges"]):
-                raise exceptions.channelModeratedException()
 
-            # Make sure we are in the channel
-            if recipient_name not in await osuToken.get_joined_channels(
-                userToken["token_id"],
-            ):
-                raise exceptions.userNotInChannelException()
+async def enqueue_public_message(
+    user_token: osuToken.Token,
+    channel_name: str,
+    client_channel_name: str,
+    message: str,
+    but: list[str] = [],
```

Comment:
Watch out for using mutable function defaults in python -- it's one of the biggest antipatterns/traps because the initialization of the default value is constructed at the time of _function definition rather than function invocation_

This means:
```py
>>> def f(x = []):
...   x.append(1)
...   return x
...
>>> print(f())
[1]
>>> print(f())
[1, 1]
```